### PR TITLE
Add setMailboxes#onDestroyRemoveMessages argument

### LINKED ISSUE #69: draft-ietf-jmap-mail: Destroying mailboxes and how this affects messages within them
2.3.4.  Destroying mailboxes

   A mailbox MUST NOT be destroyed if it has any messages assigned to
   it.  Attempts to do so MUST be rejected with a "mailboxHasMessage"
   error.

Previous version of the document required for messages to be moved to INBOX if they were in another mailbox and that was the only mailbox for them. Arguably the above is better, but it is still a departure from IMAP model, which causes all messages to be destroyed (unless they are also referenced by another JMAP mailbox). I think this needs to be further discussed in the WG.

### PR DESCRIPTION
**Pull request by neilj**
Resolves #69, see mailing list discussion.

### CODE REVIEW

**Comment by neilj** [comment]
> I think of a corner case: what if the onLastReferenceMoveTo mailbox do not exist?

This is already covered in the proposed change; the setMailboxes call will be rejected with an invalidArguments error.

> What if I want to specify by role instead? IE "Go to trash?" (Because as far as I remember the "trash" folder can change, for instance.)

I'm not convinced this is worth the extra complexity; you can use the *ifInState* parameter if you're really concerned about mailbox changes since your last sync.

**Comment by neilj** [comment]
> Also, how does this interact with IMAP?

This has no effect on IMAP, which will continue to behave exactly as before.

> Moreover, is it a mailbox property? Or an "account level property"?

Neither, this is an argument to the setMailboxes method; it is not stored anywhere.

**Comment by chibenwa** [comment]
Ok, got it. It makes sens to have it only when I destroy a mailbox.

File: spec/mail/mailbox.mdown [hunk 1]
```diff
@@ -221,7 +232,7 @@ If the mailbox has `mayDelete == false`, the destruction MUST be rejected with a
 
 A mailbox MUST NOT be destroyed if it still has any child mailboxes. Attempts to do so MUST be rejected with a `mailboxHasChild` error.
 
-A mailbox MUST NOT be destroyed if it has any messages assigned to it. Attempts to do so MUST be rejected with a `mailboxHasMessage` error.
+Before destroying the mailbox, all messages currently assigned to it must be removed from the mailbox. If a message is not in any other mailboxes, it MUST be added to the mailbox given by the *onLastReferenceMoveTo* argument, if given, otherwise destroyed.
```
Comment (chibenwa) [inline spec/mail/mailbox.mdown]:
```
Before destroying the mailbox, all messages currently assigned to it must be removed from the mailbox.
```

Should rather be:

```
Before destroying the mailbox, all messages currently assigned to it SHOULD be removed from the mailbox.
```

Comment (neilj) [reply spec/mail/mailbox.mdown]: No, it can't be a SHOULD; what else would the server do? It's an interoperability issue if it's not a must.

Comment (arnt) [inline spec/mail/mailbox.mdown]: Shouldn't "given" be "non-null"?

If not supplying the argument triggers implicit destruction of other messages, then the default is against JMAP's principles. Having that destruction happen if the client supplies an explicit null seems more acceptable.

File: spec/mail/mailbox.mdown [hunk 2]
```diff
@@ -160,6 +160,17 @@ Mailboxes can be created, updated and destroyed using the *setMailboxes* method.
   A map of mailbox id to objects containing the properties to update for that Mailbox. If `null`, no objects will be updated.
 - **destroy**: `String[]|null`
   A list of ids for Mailboxes to permanently delete. If `null`, no objects will be deleted.
+- **onLastReferenceMoveTo**: `String|null`
+  A message must be in at least one mailbox at all times. Destroying mailboxes will remove any messages in them from that mailbox. If the message is not in any other mailboxes, it will be added to the mailbox with the id given by this argument, if not `null`. If `null`, the message will be destroyed.
```
Comment (chibenwa) [inline spec/mail/mailbox.mdown]: A remark triggered internally, regarding JMAP philosophy by https://github.com/fvn-linagora :

```
not providing a value to onLastReferenceMoveTo (eqv to set as null) should NOT trigger any destructive operation.
```

That is, the mailbox SHOULD NOT be removed and the operation SHOULD be rejected with a `mailboxHasMessage` error.
```

**Comment by jeffpc** [comment]
Nice & simple, thanks.

**Pull request merged**